feat(ot3): Add device info caching mechanism to keep track of submodule versions for updates.

vegano1 opened the pull request:
# Overview

We need to update the subsystems on the OT3 (head, gantry-x, gantry-y, gripper, and pipettes) in order to do so we need to keep track of the versions of these systems. This will add the caching mechanism so we can then use that to determine if we need to do an update or not.

Closes: [RCORE-476](https://opentrons.atlassian.net/browse/RCORE-476)

# Test Plan
- [x] Robot server starts up and stores the device info cache internally 
- [x] The device info is gathered when the system starts up
- [x] Make sure probe still works as intended

# Changelog
- Added NetworkInfo class to provide network interaction and keep device info cache
- Moved probe into NetworkInfo class so we can update our internal cache state
- Added DeviceInfoCache dataclass to house DeviceInfoResponse data
- Probe now returns a dict of Nodes and DeviceInfoCache 

# Review requests

# Risk assessment
Low, OT3 only

vegano1 committed:
added device_info module to handle DeviceInfoRequest and DeviceInfoResponse messages
device_info is now cached in ot3controller._device_info_cache dict
this is updated when the network is probed which happens when the ot3 starts up and when the instruments are cached

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -915,3 +922,23 @@ async def capacitive_pass(
         return data
+
+    async def update_motor_status(self) -> None:
+        """Retreieve motor and encoder status and position from all present nodes"""
```
```suggestion
        """Retreive motor and encoder status and position from all present nodes"""
```

sfoster1 commented on hardware/opentrons_hardware/hardware_control/network.py:
```diff
@@ -1,13 +1,15 @@
 import logging
-from typing import Set, Optional
+from typing import Dict, Set, Optional
+from opentrons.hardware_control.types import DeviceInfoCache
```
we've previously been keeping `opentrons_hardware` as an independent module that is consumed by `opentrons`. let's try and stick with that - it makes this module more usable for manufacturing tests, and in general it's nice to have only one direction of dependencies - and move `DeviceInfoCache` in here.

sfoster1 commented on hardware/opentrons_hardware/firmware_update/device_info.py:
```diff
@@ -0,0 +1,68 @@
+"""Utilities for gathering device info for OT3 submodules."""
```
IMO this stuff all should live in `opentrons_hardware.hardware_control.network`. It's an extension of the stuff in the network-probing infrastructure - you can consider it as doing what the network-probe does but keeping some extra information - and not really an extension of firmware update.

sfoster1 reviewed (changes requested):
I think we should move the `opentrons_hardware.firmware_update.device_info` into the network probe infrastructure. That's because
1. what's in `firmware_update.device_info` is useful for firmware update, but not at the level of opentrons_hardware - it informs higher layers of what things there are to update, but it's disjoint from the functionality presented _by opentrons_hardware_ for firmware update. 
2. It is very similar to the functionality presented by opentrons_hardware for network probing - it's network probing but storing some extra information. It uses the same messages as probe, it needs to be re-run in the same circumstances as probe, it's used in the same way as probe.
3. Once it's in the same place as probe, it can be consolidated _into_ probe, and I think you'll find that once you do that it's a pretty good idea (and easy implementation) to push down the extra logic and storage currently in `ot3controller.py` into the probe submodule.

vegano1 committed:
Merge branch 'edge' into RCORE-476_fw_version_cache

vegano1 committed:
formating

vegano1 committed:
Update api/src/opentrons/hardware_control/backends/ot3controller.py

Co-authored-by: Seth Foster <seth@opentrons.com>

vegano1 committed:
consolidate device_info module into network module
add NetworkInfo class to network module that is responsible for storing network_info about the nodes on the network
moved probe into NetworkInfo class so we can cache the response we get whenever probe is called

vegano1 committed:
Merge branch 'RCORE-476_fw_version_cache' of https://github.com/Opentrons/opentrons into RCORE-476_fw_version_cache

vegano1 committed:
fix probe since it was not returning all attached nodes

vegano1 committed:
lint

vegano1 committed:
formatting...

sfoster1 commented on hardware/opentrons_hardware/hardware_control/network.py:
```diff
@@ -1,71 +1,117 @@
-    """Probe the bus and discover connected devices.
+@dataclass
+class DeviceInfoCache:
+    node_id: int
```
should probably be a `NodeId` enum object from `firmware_bindings.constants`

sfoster1 reviewed (changes requested):
Looks good to me! Really like the combined functionality. We should capture the node ids as enum members though.

vegano1 committed:
Merge branch 'edge' into RCORE-476_fw_version_cache

vegano1 committed:
cache NodeId in DeviceInfoCache instead device_id int

sfoster1 reviewed (approved):
Nice, looks good to me!
